Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp

```diff
@@ -3678,6 +3682,30 @@ std::pair<B3::PatchpointValue*, PatchpointExceptionHandle> AirIRGenerator::emitC
     return { patchpoint, exceptionHandle };
 }
 
+std::pair<B3::PatchpointValue*, PatchpointExceptionHandle> AirIRGenerator::createReturnCallPatchpoint(BasicBlock* block, Checked<int32_t>& newStackOffset, Vector<ArgumentLocation>& locations, const Vector<TypedTmp>& args, Vector<ConstrainedTmp> extraArgs)
+{
+    auto* patchpoint = addPatchpoint(B3::Void);
+    patchpoint->effects.terminal = true;
+    patchpoint->clobberEarly(RegisterSet::macroScratchRegisters());
+    patchpoint->clobberLate(RegisterSet::volatileRegistersForJSCall());
+    RegisterSet calleeSaveRegisters = RegisterSet::calleeSaveRegisters();
+    calleeSaveRegisters.exclude(RegisterSet::stackRegisters());
+    patchpoint->clobber(calleeSaveRegisters);
+
+    // NOTE: We reuse the caller stack space and spill the exceeding arguments to avoid corrupting slots below the frame pointer. They will be moved to the correct locations in the generator.
+    for (unsigned i = 0; i < args.size(); ++i) {
+        Tmp tmp = args[i];
+        if (locations[i].isStackArgument() && newStackOffset + locations[i].offsetFromSP() > CallFrame::returnPCOffset()) {
+            append(block, tmp.isGP() ? Move : MoveDouble, tmp, Arg::addr(Tmp(GPRInfo::callFrameRegister), newStackOffset + locations[i].offsetFromSP()));
+            continue;
+        }
+        extraArgs.append(ConstrainedTmp(tmp, locations[i]));
+    }
+    PatchpointExceptionHandle exceptionHandle = preparePatchpointForExceptions(patchpoint, extraArgs);
+    emitPatchpoint(block, patchpoint, Tmp { }, WTFMove(extraArgs));
```

Comment:
Can we get some tests for exceptions? I am not sure how this is supposed to work, since the spec doesn't mention it at all. I would assume that an exception thrown by the callee would propagate directly to the caller's caller, since there is no stack frame for the caller anymore? Call_indirect itself cannot throw, so this shouldn't be needed, right?

Reply:
I will add some tests, but I would assume the same yes. I'll need to take a closer look to the exceptions implementation.